### PR TITLE
Improve pin/unpin in MutArray

### PR DESCRIPTION
**Pull request by adithyaov**

### CODE REVIEW

File: core/src/Streamly/Internal/Data/MutArray/Type.hs [hunk 1]
```diff
@@ -2086,24 +2092,30 @@ fromListRev :: (MonadIO m, Unbox a) => [a] -> m (MutArray a)
...
+    liftIO $ do
+        let startSrc = arrStart src
+            srcLen = arrEnd src - startSrc
+        newArrContents <- Unboxed.newBytesAs ps srcLen
+        putSliceUnsafe (arrContents src) startSrc newArrContents 0 srcLen
+        return $ MutArray newArrContents 0 srcLen srcLen
+
+{-# INLINE clone #-}
+clone :: MonadIO m => MutArray a -> m (MutArray a)
+clone = cloneAs Unpinned
```
Comment (harendra-kumar) [inline core/src/Streamly/Internal/Data/MutArray/Type.hs:2110]: Is this operation faster than streaming the array and writing it to a new array? Why do we need this?

File: core/src/Streamly/Internal/Data/MutArray/Type.hs [hunk 2]
```diff
@@ -242,6 +245,7 @@ import Streamly.Internal.Data.Unbox
     , PinnedState(..)
     , getMutableByteArray#
     , touch
+    , putSliceUnsafe
```
[the remark this answers is not in the document]
Comment (adithyaov) [reply core/src/Streamly/Internal/Data/MutArray/Type.hs:248]: Can remove this.

File: core/src/Streamly/Internal/Data/MutArray/Type.hs (hunk 1 above)
Comment (adithyaov) [reply core/src/Streamly/Internal/Data/MutArray/Type.hs:2110]: I'll need to check if it's faster.

File: core/src/Streamly/Internal/Data/MutArray/Type.hs (hunk 2 above)
[the remark this answers is not in the document]
Comment (adithyaov) [reply core/src/Streamly/Internal/Data/MutArray/Type.hs:248]: Wait. Cannot remove this?

File: core/src/Streamly/Internal/Data/MutArray/Type.hs (hunk 1 above)
Comment (adithyaov) [reply core/src/Streamly/Internal/Data/MutArray/Type.hs:2110]: See https://github.com/composewell/streamly/issues/2573

**Pull request merged**